src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp

```diff
@@ -0,0 +1,127 @@
+                            bool freeK) :
+        ElectronicID {std::make_unique<pcsc_cpp::SmartCard>()},
+        certContext {certCtx}, certData {cert}, certType {cType},
+        // TODO: SignatureAlgorithm::PS?
+        signatureAlgo {isRsa ? SignatureAlgorithm::RS : SignatureAlgorithm::ES}, key {k}, freeKey {
+                                                                                              freeK}
+    {
+    }
+
+    virtual ~MsCryptoApiElectronicID()
+    {
+        if (freeKey) {
+            NCryptFreeObject(key);
+        }
+        CertFreeCertificateContext(certContext);
+    }
+
+    // The following placeholders are not used as the external PIN dialog manages PIN length
+    // validation.
+    static const int8_t PIN_RETRY_COUNT_PLACEHOLDER = -1;
+    static const size_t PIN_LENGTH_PLACEHOLDER = 0;
+
+private:
+    // Use the external dialog provided by the CryptoAPI cryptographic service provider.
+    bool providesExternalPinDialog() const override { return true; }
+
+    pcsc_cpp::byte_vector getCertificate(const CertificateType typ) const override
+    {
+        if (typ != certType) {
+            THROW(WrongCertificateTypeError,
+                  "This electronic ID does not support " + std::string(typ) + " certificates");
+        }
+        return certData;
+    }
+
+    JsonWebSignatureAlgorithm authSignatureAlgorithm() const override;
+
+    PinMinMaxLength authPinMinMaxLength() const override
+    {
+        return {PIN_LENGTH_PLACEHOLDER, PIN_LENGTH_PLACEHOLDER};
+    }
+
+    PinRetriesRemainingAndMax authPinRetriesLeft() const override
+    {
+        return {uint8_t(PIN_RETRY_COUNT_PLACEHOLDER), PIN_RETRY_COUNT_PLACEHOLDER};
+    }
+
+    pcsc_cpp::byte_vector signWithAuthKey(const pcsc_cpp::byte_vector& pin,
+                                          const pcsc_cpp::byte_vector& hash) const override;
+
+    const std::set<SignatureAlgorithm>& supportedSigningAlgorithms() const override
+    {
+        return isRSA() ? RSA_SIGNATURE_ALGOS() : ELLIPTIC_CURVE_SIGNATURE_ALGOS();
+    }
+
+    PinMinMaxLength signingPinMinMaxLength() const override
+    {
+        return {PIN_LENGTH_PLACEHOLDER, PIN_LENGTH_PLACEHOLDER};
+    }
+
+    PinRetriesRemainingAndMax signingPinRetriesLeft() const override
+    {
+        return {uint8_t(PIN_RETRY_COUNT_PLACEHOLDER), PIN_RETRY_COUNT_PLACEHOLDER};
+    }
+
+    Signature signWithSigningKey(const pcsc_cpp::byte_vector& pin,
+                                 const pcsc_cpp::byte_vector& hash,
+                                 const HashAlgorithm hashAlgo) const override;
+
+    std::string name() const override
+    {
+        // TODO: use NCryptGetProperty(key, NCRYPT_NAME_PROPERTY, ...).
+        return "MS CryptoAPI electronic ID";
+    }
+    Type type() const override { return Type::MsCryptoApiEID; }
+
+    bool isRSA() const { return signatureAlgo != SignatureAlgorithm::ES; }
+
+    const PCCERT_CONTEXT certContext;
```

Comment:
```suggestion
   PCCERT_CONTEXT certContext;
```
Type name indicates already const

Reply:
Thanks, fixed.